src/te/operation/cross_thread_reduction.cc

```diff
@@ -52,9 +99,19 @@ Stmt MakeCrossThreadReduction(
     CHECK(reduce);
     reduces[i] = reduce;
   }
-  PrimExpr cond = reduces[0]->condition;
-  for (PrimExpr v : conds) {
-    cond = cond && v;
+
+  // This computes the bound checking predicates in normal reduction.
+  auto normal_conds = MakeBoundCheck(stage, dom_map, value_map, false,
+    std::unordered_set<IterVar>());
+
+  // The existing reduction predicate (only from the first one one?)
+  PrimExpr reduction_pred = reduces[0]->condition;
+
+  // normal_pred = reduction_pred && normal_pred
```

Comment:
In my understanding, normal reduction is performed in many input threads, which is guarded by `normal_pred`, while cross-thread reduction then reduces them into one thread, which is guarded by `reduction_pred`. Therefore, should `reduction_pred` be stricter than `normal_pred`? Should it be `reduction_pred = reduction_pred && normal_red`?

Reply:
My notation is:

-   reduction_pred = input predicate from input, rarely used though
-   normal_pred = bound checking predicate

Inside the normal reduction loop, both predicates are needed. Outside the loop, only reduction_pred will be applied.  The existing logic applies both predicates inside and outside loop, which is wrong.

Reply:
I think it better to rename `reduction_pred` to `input_pred`, and `thread_head_check` to `output_pred`.

Reply:
sounds good to me. let me update the patch.